### ISSUE TITLE
SEC-41:  EPIC:   Use PKCS as the default keystore

### ISSUE THREAD

**Issue by emilytee** [OpenLiberty/open-liberty#6691]

**Comment by emilytee** [comment]
https://apps.na.collabserv.com/wikis/home?lang=en-us#!/wiki/Wde4b45b770c2_47cd_9d6d_7061f359add9/page/Feature%20Release%20Readiness%20Checklist%20Template

**Comment by marikaj123** [comment]
@emilytee - Please provide a link to the AHA featue.

**Comment by marikaj123** [comment]
@emilytee - Does the feature has files that require translation?

**Comment by skasund** [comment]
L2 has requested STE slides for this feature.  The STE template can be found at the link below: 

https://ibm.box.com/s/1an42g7zdgmaj84w7dft0indqfgi8ffm

Please upload the completed slides to the same STE Archive BOX folder.  Thanks!

**Comment by emilytee** [comment]
Alasdair added his no_design_approved tag to  https://bigblue.aha.io/features/SEC-41

**Comment by gscottj** [comment]
Accessibility Approval Comment:  
Based on the description provided, it appears that this feature does not have any end-user user interfaces that require accessibility compliance testing. End user user-interfaces include GUI, CLI and documentation that users are expected to interact with to accomplish tasks or goals. If this assumption is not correct, then let me know that required accessibility verification testing of the UI will be performed before it is shipped and what are the results from those tests.

**Comment by jhanders34** [comment]
Startup CPU increases with this features, but the security team explained that is because PKCS requires additional processing.  Startup time is only marginally affected.

**Comment by jhanders34** [comment]
Startup CPU increases with this features, but the security team explained that is because PKCS requires additional processing.  Startup time is only marginally affected.

**Comment by marikaj123** [comment]
Approving since all translated messages are in Master.

**Comment by sabolo** [comment]
Please can you complete the feature test review for this epic in order for the FAT review to progress.  I have created the child issue and added the template for you.

**Comment by emilytee** [comment]
Link to AHA feature:  https://bigblue.aha.io/features/SEC-41

**Comment by atosak** [comment]
@donbourne , what do we need to provide for serviceability?

**Comment by donbourne** [comment]
@emilytee can you please describe what changes from end user perspective for this feature?  not sure if usual serviceability approval questions apply yet.

**Comment by donbourne** [comment]
@emilytee not sure if this applies to this feature yet (see above comment) but just in case, here are the usual serviceability approval questions:

**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation?

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature?
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  

4.  Which L2 / L3 queues will handle PMRs for this feature?  Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

**Comment by emilytee** [comment]
FAT testing - feature test summary:

A new FAT bucket has been introduced in Open Liberty, com.ibm.ws.ssl_fat.pkcs12 which tests the following:

FAT Test Scenarios:

    With no keystore config at all in server.xml, make sure that a PKCS12 keystore gets created on start-up

    DefaultJKSDoesNotExistSSLTest
    testSSLFeatureNoSSLConfigNoExistingJKSGensPKCS12KeyStore

    With no keystore config at all in server.xml, and with a key.jks keystore that exists in the default location, make sure that the existing key.jks is used and no PKCS12 keystore is created

    DefaultJKSExistsSSLTest
    testSSLFeatureNoSSLConfigExistingJKSUsesJKSKeyStore

    With a a minimal SSL configuration, e.g., , and a key.jks keystore file that exists in the default location, make sure the default key.jks keystore
    is used and no PKCS12 keystore is created.

    DefaultJKSExistsSSLTest
    testDefaultMinimalSSLCertificateFileKeyJKSFileDoesExist

    With a a minimal SSL configuration, e.g., , and a key.p12 keystore file that exists in the default location, make sure the default key.p12 keystore
    is used

    DefaultPKCS12ExistsSSLTest
    testDefaultMinimalSSLCertificateFileKeyPKCS12FileDoesExist

    Make sure a new JKS keystore will be created if there is a minimal SSL configuration where a type is specified and the type is JKS
    e.g.,

    DefaultJKSDoesNotExistSSLTest
    testDefaultMinimalSSLConfigWithJKSType

    Make sure a new PKCS12 keystore will be created if there is a minimal SSL configuration where a type is specified and the type is PKCS12
    e.g.,

    DefaultPKCS12DoesNotExistSSLTest
    testDefaultMinimalSSLConfigWithPKCS12Type

    When a keystore is specified in server.xml and it is not the default keystore, and the type is not specified, but the location of the
    JKS keytore is specified, that JKS keystore will still be used. Verify this scenario will have no effect on the default keystore behavior.
    e.g.,

    NonDefaultJKSSSLTest
    testNonDefaultSSLConfigUsingJKS

    When a keystore is specified in server.xml and it is not the default keystore, and the type is not specified, but the location of the
    PKCS12 keytore is specified, that PKCS12 keystore will still be used. Verify this scenario will have no effect on the default keystore behavior.
    e.g.,

    NonDefaultPKCS12SSLTest
    testNonDefaultSSLConfigUsingPKCS12

**Comment by emilytee** [comment]
For serviceability (@donbourne )

Serviceability Approval Comment - Please answer the following questions for serviceability approval:

    WAD -- does the WAD identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR? Have these issues been addressed in the implementation?

==>  No WAD was needed.    Users will not see anything different in their environment.  If they have specified a default keystore configuration in their server.xml, we will now be creating a PKCS12 keystore file instead of the JKS keystore file we had been.    The messages.log will contain the name and type of the default keystore created.     

    Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).
    a) What problem paths were tested and demonstrated?  (see the FAT content comment earlier)
    b) Who did you demo to?   (Security team, SVT)
    c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3?   Yes.

    SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them. Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
    a) Who conducted SVT tests for this feature?    Brian H
    b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?
Yes, they see the error issues being emitted to the messages log and those errors are sufficient to diagnose the problem.

    Which L2 / L3 queues will handle PMRs for this feature? Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it. Ask Don Bourne if you need links or more info.     WASSEC queue will handle PMRs for this.

**Comment by emilytee** [comment]
Feature Test Summary Part 2:

Part 2:
Collectively as a team you need to assess your confidence in the testing delivered based on the values below. This should be done as a team and not an individual to ensure more eyes are on it and that pressures to deliver quickly are absorbed by the team as a whole.

    Please indicate your confidence in the testing (up to and including FAT) delivered with this feature by selecting one of these values:
    0 - No automated testing delivered
    1 - We have minimal automated coverage of the feature including golden paths. There is a relatively high risk that defects or issues could be found in this feature.
    2 - We have delivered a reasonable automated coverage of the golden paths of this feature but are aware of gaps and extra testing that could be done here. Error/outlying scenarios are not really covered. There are likely risks that issues may exist in the golden paths
    3 - We have delivered all automated testing we believe is needed for the golden paths of this feature and minimal coverage of the error/outlying scenarios. There is a risk when the feature is used outside the golden paths however we are confident on the golden path. Note: This may still be a valid end state for a feature... things like Beta features may well suffice at this level.
    4 - We have delivered all automated testing we believe is needed for the golden paths of this feature and have good coverage of the error/outlying scenarios. While more testing of the error/outlying scenarios could be added we believe there is minimal risk here and the cost of providing these is considered higher than the benefit they would provide.
    5 - We have delivered all automated testing we believe is needed for this feature. The testing covers all golden path cases as well as all the error/outlying scenarios that make sense. We are not aware of any gaps in the testing at this time. No manual testing is required to verify this feature.

    Based on your answer above, for any answer other than a 4 or 5 please provide details of what drove your answer. Please be aware, it may be perfectly reasonable in some scenarios to deliver with any value above. We may accept no automated testing is needed for some features, we may be happy with low levels of testing on samples for instance so please don't feel the need to drive to a 5. We need your honest assessment as a team and the reasoning for why you believe shipping at that level is valid. What are the gaps, what is the risk etc. Please also provide links to the follow on work that is needed to close the gaps (should you deem it needed)

Confidence:  5
Comments:  We, as a team, have reviewed and identified all of the golden paths as well as the error paths, and have included automated testing for each of those paths.

**Comment by dave-waddling** [comment]
@emilytee Thanks for the above, FAT Focal Approval granted!

**Comment by emilytee** [comment]
Epic is included in 19.0.0.3 GM `Liberty Official Release Build - Static cl190320190321-1636`

https://wasrtc.hursley.ibm.com:9443/jazz/resource/itemOid/com.ibm.team.build.BuildResult/_Hf5B6UvvEemPWpVfp3aH-ghttps://libfsfe01.hursley.ibm.com/liberty/dev/Xo/release/cl190320190321-1636-_Hf5B6UvvEemPWpVfp3aH-g/fe/cl190320190321-1636.16.linux//linux/zipper/externals/installableshttps://rtpgsa.ibm.com/projects/l/liberty.build/liberty/WASX.LIBERTY/daily//cl190320190321-1636.16.linux//linux/zipper/externals/installables (edited)
